test/mlir/onnx/onnx_lowering.mlir

```diff
@@ -2139,3 +2139,23 @@ func @test_lstm_bidirectional_mode(%arg0: tensor<4x3x2xf32>, %arg1: tensor<1x12x
   // CHECK:  %[[REVERSE_SEQUENCE_IV:.+]] = affine.apply [[REVERSE_IV_MAP]](%arg3)[%[[SEQUENCE_LEN]]{{]}}
   // CHECK:  [[Xt_LOAD:%.+]] = load %arg0[%[[REVERSE_SEQUENCE_IV]], {{.*}}, {{.*}}] : memref<4x3x2xf32>
 }
+
```

Comment:
In the code, when you lower, you have 2 cases: one with hasAllConstantDimensions and one without. Would it make sense to check both? When we have constants, why are we not generating the size as a single constant? I know LLVM would catch it but it may make later opt confused?

Also on the lowering, what happen if the input of squeeze is used only once? Can we have a dummy "squeeze in place" that simply do nothing? Or would we be able to simply remove squeeze (but I think it would confuse shape inference) So we may still need a "squeeze with a <in-place> parameter, and then the lowering of this becomes a no-op?

Reply:
@AlexandreEichenberger 

I updated the code as follows.
- if hasAllConstantDimensions, emit the tensor size as a single constant,
- otherwise, try to collect all constant dimensions and emit a single constant for them, then multiple it with dynamic dimensions.
- Added a lit test for the latter case.

>Also on the lowering, what happen if the input of squeeze is used only once? Can we have a dummy "squeeze in place" that simply do nothing?
Could you elaborate more on this? If the input of squeeze is used only once, we still need to have its output in a correct shape. I am not sure whether it is possible to simply do nothing.

Reply:
@tungld 

My understanding is that removing dimensions that has a size of 1 does not impact the actual code.

int a[10][1][2];
int b[10][2];

a[5][0[1] is &a + 5*(10*1*2)+ 0*(2) + 1 *(1)
b[5][1] is &a + 5*(10*2) + 1 * (1)

both a & b array reference have the same offset with respect to their base address. So it should become a no-op

Reply:
Thanks! I'll look at how to remove a dimension in a tensor/memref without introducing memcpy. By the way, do you think it is good to do such an elimination here in the lowering code or in the canonicalization pass (We eliminated identity ops in the canonicalization pass)?

Reply:
@AlexandreEichenberger 

I added a special case where the input is used only once, as you suggestion. It is done by using `krnl.getref` which does not allocate additional memory but creates a new view of the input.

Thanks!